clients/src/main/java/org/apache/kafka/server/policy/TopicPolicy.java

```diff
@@ -0,0 +1,214 @@
+        public String name() {
+            return name;
+        }
+
+        /**
+         * The current number of partitions.
+         * This will be 0 for a {@link Type#CREATE_TOPIC} request.
+         */
+        public int currentPartitions() {
+            return currentPartitions;
+        }
+
+        /**
+         * The requested number of partitions. The optional will be empty if a user defined replica assignment was supplied.
+         * This will be greater than {@link #currentPartitions()} for a {@link Type#CREATE_TOPIC} or
+         * {@link Type#CREATE_PARTITIONS} request.
+         */
+        public Optional<Integer> requestedPartitions() {
+            return requestedPartitions;
+        }
+
+        /**
+         * The current replication factor,
+         * defined as the number of replicas of the partition with the smallest partition index.
+         * This will be 0 for a {@link Type#CREATE_TOPIC} request.
+         */
+        public short currentReplicationFactor() {
+            return currentReplicationFactor;
+        }
+
+        /**
+         * The requested replication factor. This optional will be empty if a user defined replica assignment was supplied.
+         * This will be greater than {@link #currentPartitions()} for a {@link Type#CREATE_TOPIC} or
+         * {@link Type#REASSIGN_PARTITIONS} request.
+         */
+        public Optional<Short> requestedReplicationFactor() {
+            return requestedReplicationFactor;
+        }
+
+        /**
+         * The replica reassignments for this topic. This is in the form of a map from partition id to replica (broker) ids.
+         * This will be empty for anything other than {@link Type#CREATE_TOPIC} and {@link Type#REASSIGN_PARTITIONS} requests.
+         * For a {@link Type#CREATE_TOPIC} event this will be either populated with the user supplied replica assignment
+         * (in which case requestedPartitions and requestedReplicationFactor will be empty) the auto-generate assignment
+         * if partition count and/or replication factor were supplied.
+         */
+        public Map<Integer, List<Integer>> replicaAssignments() {
+            return replicaAssignments;
+        }
+
+        /**
+         * The current configs of the topic.
+         * This will be empty for a {@link Type#CREATE_TOPIC} request.
+         */
+        public Map<String, String> currentConfigs() {
+            return currentConfigs;
+        }
+
+        /**
+         * The requested configs of the topic.
+         * This will be different from {@link #currentConfigs()} for a {@link Type#CREATE_TOPIC} or
+         * {@link Type#UPDATE_CONFIGS} request.
+         */
+        public Map<String, String> requestedConfigs() {
+            return requestedConfigs;
+        }
+    }
+
+    enum Type {
+        CREATE_TOPIC,
+        CREATE_PARTITIONS,
+        REASSIGN_PARTITIONS,
+        UPDATE_CONFIGS,
+        DELETE_TOPICS
+    }
+
+    class RequestMetadata {
```

Comment:
I think we want the principal too

Reply:
yes the `KafkaPrincipal` that initiated the request

Reply:
Happy to add that, the current `CreateTopicPolicy` doesn't get that information though.

Reply:
Ok, I have added the `KafkaPrincipal` to the `TopicPolicy.validate` method. I had to make it `Optional` as there seem to be some cases (such as in auto topic creation) where the principal could be null (because the context object is null).

Reply:
Another option could be to use `KafkaPrincipal.ANONYMOUS`. Are there many cases where we don't currently have the principal? Even for auto create, it should always be triggered by a client request that should have a principal.